Faster hmac sha1 token generation

nhinds opened the pull request:
Switch from the [jsSHA](https://github.com/Caligatio/jsSHA) library to [js-sha1](https://github.com/emn178/js-sha1), which is roughly twice as fast on my watch. `js-sha1` doesn't have HMAC support built in, so that's now a custom implementation in `common/hmac-sha1.js`. Using `Uint8Array` everywhere instead of encoding and decoding hex/base32 strings gives a significant performance boost while computing tokens.

To avoid decoding the tokens from base32 each time the app launches, this PR changes the stored CBOR file from storing tokens as base32-encoded strings to storing the decoded byte arrays. Existing stored tokens will be converted when the app is first launched, and new tokens will be converted when they are added in settings. It would be possible to re-encode the byte arrays back to base32 if you want to switch out the libraries for OTP/HMAC/SHA1/etc. in the future.

For me, when I have 3 codes the current app takes around 4s to initially launch, and 2.5s to regenerate codes (tested by turning the screen off and on). This branch brings that to around 2.5s for the initial launch and less than 1s to regenerate codes.

nhinds committed:
Switch to js-sha1 and custom HMAC-SHA1 implementation for performance boost

Roughly doubles the performance of the HMAC-SHA1 part of TOTP generation.

nhinds committed:
Store tokens as decoded byte arrays

This avoids spending ~50-100ms converting from base32 to the
bytes used for the TOTP HMAC-SHA1 every time a code is generated.

Existing stored tokens will be converted when the app is first
launched, and new tokens will be converted when they are
added in settings.

nhinds commented:
Hi @Lixxia, is this something you would be interested in merging? Is there anything else you would like me to change / explain as part of this pull request?

Lixxia commented:
Hiya, sorry been on vacation past few weeks. This looks great! I'll take a look this weekend and run some tests. Maybe you could give me a quick overview of which situations you tested?

nhinds commented:
I tested starting with a couple of existing codes from AWS (which seems to have longer than usual secrets) and one shorter code, and those got picked up and turned into the array buffer format. I also removed them and added the same codes back from the settings page to test that saving new codes is okay.

I've been using this version of the app myself for a couple of weeks and the codes have been accepted, so at least for AWS the codes are still being generated correctly.

Lixxia commented on common/totp.js:
```diff
@@ -1,89 +1,26 @@
-    
-    let hex = parsehex(bits);
-    
-    if (hex.length % 2 !== 0) {
```
You'll need to add support for the hexpad function/check or an alternative in your new `base32ToUint8Array` function. With your current proposed changes 26 or other odd length tokens no longer work. This issue (#3) was previously solved with the addition of this function.

nhinds replied:
Whoops, sorry. I thought I had covered that, since I had been testing with Amazon tokens, but tokens with 26 characters are definitely broken with my changes. I'll get that sorted and update the branch.

nhinds replied:
Turns out just removing the check for leftover bits allows base32 strings with weird lengths to work. Tested with the 26 char test string from `app/index.js` and some other random-length base32 strings and it seems to decode properly now.

Lixxia commented on app/tokens.js:
```diff
@@ -8,6 +9,27 @@ export function AuthToken() {
+}
+
+// Convert any existing string/base32-encoded tokens to a Uint8Array, and wrap raw ArrayBuffer tokens
+function convertTokensToUint8Array(data) {
```
Given that this isn't a utility function and you're only operating on the `this.file.data` object I'd prefer that this function be integrated into the AuthToken class.

Lixxia commented:
Overall looks good, very solid approach and definitely reduces speeds on load/token regeneration. I've added a few comments in-line regarding a reintroduced bug and some code style preferences. Let me know if you have any questions!

nhinds commented:
Re-tested by removing all codes, reverting to `master`, loading in a bunch of codes (including a normal-sized code, AWS codes, and the 26-character test code from `app/index.js`), and switching back over to this branch. All the codes got migrated over to uint8array format.

Also tested removing and re-adding the same codes as above, which seemed to work fine.

Lixxia commented:
Hey, not sure if you're still looking at this or not. I did a fresh clone and install of your branch and seems there's still a few issues. There's a leftover reference to the `nodeWrap` (`p`) function in js-sha1.js that causes a `ReferenceError`. A quick attempt at trying to remove it and re-minify causes some other token handling errors down the road. Mind taking another look?

nhinds commented:
It has been working fine for me for months after the latest push, not sure what would be causing reference errors. The nodeWrap/p function should never be called in Fitbit because it's guarded by a check for node.js, which should not be true for Fitbit's js interpreter... Is something broken or are there logs somewhere that show the errors? 

I'll take a closer look later this week when I have time to debug.

nhinds committed:
Disable node.js in js-sha1 to allow the simulator to work

The simulator's companion app appears to be written in node.js,
or at least triggers js-sha1's node.js detection.

nhinds commented:
I found what you were referring to. It worked fine on my watch and phone, but I got this error when running under the simulator:
```
[8:23:36 PM] Companion: ReferenceError: p is not defined
  c at common/js-sha1.js:16,925
  ? at common/js-sha1.js:16,5266
  ? at common/js-sha1.js:16,2
```

Looks like the simulator's companion is running under node.js, or at least something that looks like node.js to the js-sha1 library. I've added the settings which disable node.js/commonjs support in js-sha1 and it seems to work fine in the simulator now. Are you able to check whether it is working for you as well?

Lixxia commented:
Everything looks good, thanks so much for your work!
